<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-проекта" data-toc-modified-id="Описание-проекта-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание проекта</a></span></li><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Заполнение-пропусков" data-toc-modified-id="Заполнение-пропусков-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Заполнение пропусков</a></span></li><li><span><a href="#Проверка-данных-на-аномалии-и-исправления." data-toc-modified-id="Проверка-данных-на-аномалии-и-исправления.-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка данных на аномалии и исправления.</a></span><ul class="toc-item"><li><span><a href="#Для-колокнки-days_employed" data-toc-modified-id="Для-колокнки-days_employed-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Для колокнки <code>days_employed</code></a></span></li><li><span><a href="#Для-колокнки-children" data-toc-modified-id="Для-колокнки-children-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Для колокнки <code>children</code></a></span></li><li><span><a href="#Для-колокнки-dob_years" data-toc-modified-id="Для-колокнки-dob_years-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Для колокнки <code>dob_years</code></a></span></li><li><span><a href="#Для-колокнки-education,-family_status,-gender" data-toc-modified-id="Для-колокнки-education,-family_status,-gender-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Для колокнки <code>education</code>, <code>family_status</code>, <code>gender</code></a></span></li></ul></li><li><span><a href="#Изменение-типов-данных." data-toc-modified-id="Изменение-типов-данных.-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Изменение типов данных.</a></span></li><li><span><a href="#Удаление-дубликатов." data-toc-modified-id="Удаление-дубликатов.-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Удаление дубликатов.</a></span></li><li><span><a href="#Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма." data-toc-modified-id="Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма.-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.</a></span><ul class="toc-item"><li><span><a href="#Создание-словаря-уровень-образования" data-toc-modified-id="Создание-словаря-уровень-образования-7.1"><span class="toc-item-num">7.1&nbsp;&nbsp;</span>Создание словаря уровень образования</a></span></li><li><span><a href="#Создание-словаря-семейное-положение" data-toc-modified-id="Создание-словаря-семейное-положение-7.2"><span class="toc-item-num">7.2&nbsp;&nbsp;</span>Создание словаря семейное положение</a></span></li><li><span><a href="#Правка-исходного-датафрейма" data-toc-modified-id="Правка-исходного-датафрейма-7.3"><span class="toc-item-num">7.3&nbsp;&nbsp;</span>Правка исходного датафрейма</a></span></li></ul></li><li><span><a href="#Категоризация-дохода." data-toc-modified-id="Категоризация-дохода.-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Категоризация дохода.</a></span></li><li><span><a href="#Категоризация-целей-кредита." data-toc-modified-id="Категоризация-целей-кредита.-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Категоризация целей кредита.</a></span></li><li><span><a href="#Ответы-на-вопросы." data-toc-modified-id="Ответы-на-вопросы.-10"><span class="toc-item-num">10&nbsp;&nbsp;</span>Ответы на вопросы.</a></span><ul class="toc-item"><li><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Вопрос-1:" data-toc-modified-id="Вопрос-1:-10.0.0.1"><span class="toc-item-num">10.0.0.1&nbsp;&nbsp;</span>Вопрос 1:</a></span></li><li><span><a href="#Вывод-1:" data-toc-modified-id="Вывод-1:-10.0.0.2"><span class="toc-item-num">10.0.0.2&nbsp;&nbsp;</span>Вывод 1:</a></span></li><li><span><a href="#Вопрос-2:" data-toc-modified-id="Вопрос-2:-10.0.0.3"><span class="toc-item-num">10.0.0.3&nbsp;&nbsp;</span>Вопрос 2:</a></span></li><li><span><a href="#Вывод-2:" data-toc-modified-id="Вывод-2:-10.0.0.4"><span class="toc-item-num">10.0.0.4&nbsp;&nbsp;</span>Вывод 2:</a></span></li><li><span><a href="#Вопрос-3:" data-toc-modified-id="Вопрос-3:-10.0.0.5"><span class="toc-item-num">10.0.0.5&nbsp;&nbsp;</span>Вопрос 3:</a></span></li><li><span><a href="#Вывод-3:" data-toc-modified-id="Вывод-3:-10.0.0.6"><span class="toc-item-num">10.0.0.6&nbsp;&nbsp;</span>Вывод 3:</a></span></li><li><span><a href="#Вопрос-4:" data-toc-modified-id="Вопрос-4:-10.0.0.7"><span class="toc-item-num">10.0.0.7&nbsp;&nbsp;</span>Вопрос 4:</a></span></li><li><span><a href="#Вывод-4:" data-toc-modified-id="Вывод-4:-10.0.0.8"><span class="toc-item-num">10.0.0.8&nbsp;&nbsp;</span>Вывод 4:</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#Общий-вывод:" data-toc-modified-id="Общий-вывод:-11"><span class="toc-item-num">11&nbsp;&nbsp;</span>Общий вывод:</a></span></li></ul></div>

# Исследование надежности заемщиков.

## Описание проекта 

**Заказчик**

Кредитный отдел банка.

**Задача**

Изучить влияние различных факторов на возврат кредита в срок.

## Обзор данных

In [1]:
import pandas as pd # импорт библиотеки pandas

In [2]:
df = pd.read_csv('datasets\data.csv') # чтение файла с данными и сохранение в df

In [3]:
df.head(10) # получение первых 10 строк таблицы df

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [4]:
df.info() # получение общей информации о данных в таблице df

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Заполнение пропусков

In [5]:
df.isna().sum() # подсчёт пропусков

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Пропущенные значени в `days_employed` и `total_income` - это данный по общему трудовому стажу в днях и ежемесячному доходу соответственно. Пропущенные значени в количественных данных.

In [6]:
round(df['days_employed'].isna().sum() / df.shape[0], 3) # доля пропусков для колонки days_employed

0.101

In [7]:
round(df['total_income'].isna().sum() / df.shape[0], 3) # доля пропусков для колонки total_income

0.101

In [8]:
df[(df['days_employed'].isna()) & (df['total_income'].notna())]['days_employed'].count()  # проверка взаимосвязи пропусками 
                                                                         # в days_employed и total_income

0

In [9]:
df[df['days_employed'].isna()].head(10)

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   
65         0            NaN         21   среднее             1   
67         0            NaN         52    высшее             0   
72         1            NaN         32    высшее             0   
82         2            NaN         50    высшее             0   
83         0            NaN         52   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   
65  Не женат / не замужем                 4      M    компаньон     0   
67        женат / замужем                 0      F    пенсионер     0   
72        женат / замужем                 0      M  госслужащий     0   
82        женат / замужем                 0      F    сотрудник     0   
83        женат / замужем                 0      M    сотрудник     0   

    total_income                                purpose  
12           NaN                        сыграть свадьбу  
26           NaN                            образование  
29           NaN       строительство жилой недвижимости  
41           NaN       сделка с подержанным автомобилем  
55           NaN                        сыграть свадьбу  
65           NaN  операции с коммерческой недвижимостью  
67           NaN                покупка жилья для семьи  
72           NaN  операции с коммерческой недвижимостью  
82           NaN                                  жилье  
83           NaN                                  жилье

10 процентов потенциальных заемщиков не имеют данные об общим трудовом стаже и ежемесячном доходе. При это пропуск в `days_employed` соответствует пропуску `total_income` и наоборот. Для этого могут быть следующие причины:
- заемщики могли целенаправленно не указать эти данные,
- заемщики не смогли подтвердить указанные данные, (например не имеют официального дохода),
- возможно какие-то данные были потеряны, т.к. тип данных в обеих колонок `float64` возможно при конвертации или некорректном заполнении.

In [10]:
days_employed_median = df['days_employed'].median() # медианное значением по столбцу days_employed
days_employed_median

-1203.369528770489

In [11]:
df['days_employed'].mean() # среднее значением по столбцу days_employed

63046.497661473615

In [12]:
total_income_median = df['total_income'].median() # медианное значением по столбцу total_income
total_income_median

145017.93753253992

In [13]:
df['total_income'].mean() # среднее значением по столбцу total_income

167422.3022081719

Медианным значением обеспечивает лучшее заполнение количественных переменных, т.к. позволяет нивелировать, в большей степени, влияния аномальных данных, например для колонки `days_employed` характерны отрицательные значения, но встречаются и выбросы, которые сильно влияют, на среднее значение которое составило 63046,5, если предположить что это дни то это около 172 лет стажа, что невозможно. 
Для колонки `total_income` разница не так велика, что скорее свидетельствует о однородности данных и отсутствии аномалий.       

In [14]:
df['days_employed'] = df['days_employed'].fillna(days_employed_median) #  заполнения пропусков медианными значениями
df['total_income'] = df['total_income'].fillna(total_income_median)

In [15]:
df.isna().sum() # подсчёт пропусков после заполнения

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

## Проверка данных на аномалии и исправления.

### Для колокнки `days_employed`

In [16]:
df['income_type'].value_counts() # количество записей по каждому типу занятости 

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
безработный            2
предприниматель        2
студент                1
в декрете              1
Name: income_type, dtype: int64

In [17]:
df[(df['days_employed'] >= 0) & (df['income_type'] != 'пенсионер')] # поиск аномальных значений в колонке days_employed 
                                                                    #среди всех типов занятости кроме 'пенсионер' 

children  days_employed  dob_years education  education_id  \
3133          1  337524.466835         31   среднее             1   
14798         0  395302.838654         45    Высшее             0   

          family_status  family_status_id gender  income_type  debt  \
3133    женат / замужем                 0      M  безработный     1   
14798  гражданский брак                 1      F  безработный     0   

        total_income                  purpose  
3133    59956.991984  покупка жилья для сдачи  
14798  202722.511368             ремонт жилью

In [18]:
df[(df['days_employed'] >= 0) & (df['income_type'] == 'пенсионер')]['days_employed'].count() # поиск аномальных значений в колонке days_employed 
                                                                                             # по типу занятости 'пенсионер'

3443

In [19]:
df[(df['days_employed'] == days_employed_median) & (df['income_type'] == 'пенсионер')]['days_employed'].count()#поиск медианных значений в колонке days_employed 
                                                                                               #по типу типов занятости 'пенсионер' 

413

Из полученных данных видно что аномальные значения (положительные более 300000) характерны только для двух типов занятости это: безработный и пенсионеры. Также видно что для 413 пенсионеров значение в колонке days_employed равно медианному значение, это значит, что до шага 2.1 там были значение NaN, и они были заменены для этой группы.

Предположительно данные банк вряд ли хранятся в днях еще с точностью до 6 знака, скорее всего в банке хранят дату начала трудовой деятельности, для выгрузки эти данные были переведены в дни путем вычитания даты начала трудовой деятельности минус дата подачи заявки с точностью до секунды, поэтому и получились отрицательные значения. Для категорий безработные и пенсионеры скорее всего  были указаны или диапазоны значений трудовой деятельности и некий признак, который при переводе трансформировались в непонятное числовое значение. Желательно изменить функцию обработки следующим образом:
- уйти от отрицательных значений,
- уйти от дробных значений и прийти к целочисленному формату,
- обеспечить корректную обработку для двух категорий, которые вызывают аномалии.

Предложение обработать данные данные из колонки `days_employed` следующим образом:
- для типу занятости 'пенсионер' и 'безработный' установить стаж 0, т.к. для пенсионеров их рабочий стаж уже не важен для банка, а у безработных он и так 0.
- для всех остальных типов уйти от отрицательных значений.

In [20]:
df.loc[df['income_type'] == 'пенсионер', 'days_employed'] = 0 # замена анамальных значений для типа занятости 'пенсионер'  

In [21]:
df.loc[df['income_type'] == 'безработный', 'days_employed'] = 0 # замена анамальных значений для типа занятости 'безработный'

In [22]:
df['days_employed'] = df['days_employed'].apply(abs) # преход к неотрецательным значением для значения стажа

In [23]:
df.head(10)

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0       0.000000         53   среднее             1   
5         0     926.185831         27    высшее             0   
6         0    2879.202052         43    высшее             0   
7         0     152.779569         50   СРЕДНЕЕ             1   
8         2    6929.865299         35    ВЫСШЕЕ             0   
9         0    2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

### Для колокнки `children`

In [24]:
df['children'].value_counts() # количество записей по каждому количества детей

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

In [25]:
children_min = df['children'].min() # минимальное значение для колонки children
children_max = df['children'].max() # максимального значения для колонки children

In [26]:
round(df[df['children'] == children_min]['children'].count() / df[df['children'] == 1]['children'].count(), 4) 
#доля минимальных аномальных значений к значениям с одним ребенком

0.0098

In [27]:
round(df[df['children'] == children_max]['children'].count() / df[df['children'] == 2]['children'].count(), 4) 
#доля максимальных аномальных значений к значениям с двумя детьми

0.037

Отрицательного количества детей не может быть, возможно при заполнении произошла опечатка, предлагаю заменить на 1, это не повлияет на дальнейшее иследование, т.к доля незначительна

Конечно возможно, что существует 76 заявителей с 20 детьми, но следущее после этого значение 5 и это всего 9 значений, при заполнении скорее всего произошла опечатка, предлагаю заменить на 2, это не повлияет на дальнейшее исследование, т.к доля незначительна

In [28]:
df.loc[df['children'] == df['children'].min(), 'children'] = 1 # замена минимальных аномальных значений
df['children'].min()

0

In [29]:
df.loc[df['children'] == df['children'].max(), 'children'] = 2 # замена максимальных аномальных значений
df['children'].max()

5

### Для колокнки `dob_years`

In [30]:
df['dob_years'].min() # минимальный возрас заявителя

0

In [31]:
df[df['dob_years'] != 0]['dob_years'].min() # минимальный возрас заявителя исключая 0 (реальный минимальный)

19

0 лет не может быть, заявитель должен быть не младше 18. предлагаю заполнить средними значением. При расчете среднего нулевые значения не учитывать. 

In [32]:
round(df[df['dob_years'] == 0]['dob_years'].count() / df.shape[0], 4) # доля аномальных возрастов

0.0047

In [33]:
dob_years_mean = int(df[df['dob_years'] != 0]['dob_years'].mean()) # средний возраст заявителей
dob_years_mean

43

In [34]:
df.loc[df['dob_years'] == 0, 'dob_years'] = dob_years_mean # замена минимальных аномальных значений
df['dob_years'].min()

19

In [35]:
df['dob_years'].max() #максимальный возрас заявителя

75

### Для колокнки `education`, `family_status`, `gender`

In [36]:
df['education'].value_counts() # количество записей по каждому типу семейного положения

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64

Видны неявные дубликаты необходимо привести к одному регистру, будет сделана на шаге 2.4

In [37]:
df['family_status'].value_counts() # количество  записей по каждому типу семейного положения

женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

значение `Не женат / не замужем` привести `не женат / не замужем`

In [38]:
df['family_status'] = df['family_status'].str.lower() # приведение к единому регистру
df['family_status'].value_counts()

женат / замужем          12380
гражданский брак          4177
не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

In [39]:
df['gender'].value_counts() # количество  записей по каждому типу пола

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

In [40]:
df[df['gender'] == 'XNA'] 

children  days_employed  dob_years            education  education_id  \
10701         0    2358.600502         24  неоконченное высшее             2   

          family_status  family_status_id gender income_type  debt  \
10701  гражданский брак                 1    XNA   компаньон     0   

        total_income               purpose  
10701  203905.157261  покупка недвижимости

Одни заявитель не определился, предлагаю оставить как есть.

Для остальных колонок аномалий не заметил, лишний код проверки убрал. 

## Изменение типов данных.

In [41]:
try:                                                      # изменение типа данных в колонке total_income, 
    df['total_income'] = df['total_income'].astype('int') # метод не орабатывает ошибки 
except:
    print('Что-то пошло не так, надо проверить данные')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  int32  
 11  purpose           21525 non-null  object 
dtypes: float64(1), int32(1), int64(5), object(5)
memory usage: 1.9+ MB


## Удаление дубликатов.

In [42]:
df['education'] = df['education'].str.lower() # приведение к единому регистру, для удаления неявных дубликатов
df['education'].value_counts()

среднее                15233
высшее                  5260
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

In [43]:
df.duplicated().sum() # подсчёт явных дубликатов

71

In [44]:
df = df.drop_duplicates().reset_index(drop=True) # удаление явных дубликатов (с удалением старых индексов и формированием новых)

In [45]:
df.duplicated().sum() # проверка на отсутствие дубликатов

0

Неявные дубликаты в колонке `education` были выявлены методом `value_counts()`, который показал, что в данных встречаются записи отличающиеся только регистром, это произошло в результате ручного ввода или если данные были агрегированы из нескольких источников.
Явные дубликат были проверены методом duplicated() по всему датасету, т.к. в датасете отсутствуют колонка с уникальными (ключевыми) значениями, по которым можно оценить  о наличии дубликата, например идентификатор заявителя, приходится искать их опираясь на весь кортеж. Возникновение дубликатов скорее всего связано с тем, что в колонках `days_employed` и `total_income` имелись данные со значением NaN.

Все остальные колонки проверил на шаге 2.2

## Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

### Создание словаря уровень образования

In [46]:
education_dict = df[['education_id', 'education']] # выделение словаря уровень образования

In [47]:
education_dict = education_dict.drop_duplicates().reset_index(drop=True) # удаление дубликатов из словаря уровень образования

In [48]:
education_dict.sort_values(by='education_id') # вывод словоря уровень образования

education_id            education
0             0               высшее
1             1              среднее
2             2  неоконченное высшее
3             3            начальное
4             4       ученая степень

### Создание словаря семейное положение

In [49]:
family_status_dict = df[['family_status_id', 'family_status']] # выделение словаря семейное положение

In [50]:
family_status_dict = family_status_dict.drop_duplicates().reset_index(drop=True) # удаление дубликатов из словаря семейное положение

In [51]:
family_status_dict.sort_values(by='family_status_id') # вывод словоря семейное положение

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  не женат / не замужем

### Правка исходного датафрейма

In [52]:
df = df.drop(columns=['family_status', 'education']) # Удаление из исходного датафрейма столбцов education и family_status
df.head(10)

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0       0.000000         53             1                 1      F   
5         0     926.185831         27             0                 1      M   
6         0    2879.202052         43             0                 0      F   
7         0     152.779569         50             1                 0      M   
8         2    6929.865299         35             0                 1      F   
9         0    2188.756445         41             1                 0      M   

  income_type  debt  total_income                     purpose  
0   сотрудник     0        253875               покупка жилья  
1   сотрудник     0        112080     приобретение автомобиля  
2   сотрудник     0        145885               покупка жилья  
3   сотрудник     0        267628  дополнительное образование  
4   пенсионер     0        158616             сыграть свадьбу  
5   компаньон     0        255763               покупка жилья  
6   компаньон     0        240525           операции с жильем  
7   сотрудник     0        135823                 образование  
8   сотрудник     0         95856       на проведение свадьбы  
9   сотрудник     0        144425     покупка жилья для семьи

## Категоризация дохода.

In [53]:
# Объявление функции total_income_category() с параметром total_income,
# которая возвращает категорию на основании полученого значения:
# 0–30000 — 'E';
# 30001–50000 — 'D';
# 50001–200000 — 'C';
# 200001–1000000 — 'B';
# 1000001 и выше — 'A'.

def total_income_category(total_income):
    try:                                    #на вход мо
        if 0 <= total_income <= 30000:
            return 'E'
        elif 30001 <= total_income <= 50000:
            return 'D'
        elif 50001 <= total_income <= 200000:
            return 'C'
        elif 200001 <= total_income <= 1000000:
            return 'B'
        elif 1000001 <= total_income:
            return 'А'
        else:
            return 'None'
    except:
        return 'тип данных не подходит'

In [54]:
df['total_income_category'] = df['total_income'].apply(total_income_category) # присвоение категорий по значением колонки total_income

In [55]:
df['total_income_category'].value_counts() # проверка

C    16016
B     5041
D      350
А       25
E       22
Name: total_income_category, dtype: int64

## Категоризация целей кредита.

In [56]:
df['purpose'].value_counts()

свадьба                                   791
на проведение свадьбы                     768
сыграть свадьбу                           765
операции с недвижимостью                  675
покупка коммерческой недвижимости         661
операции с жильем                         652
покупка жилья для сдачи                   651
операции с коммерческой недвижимостью     650
покупка жилья                             646
жилье                                     646
покупка жилья для семьи                   638
строительство собственной недвижимости    635
недвижимость                              633
операции со своей недвижимостью           627
строительство жилой недвижимости          624
покупка недвижимости                      621
покупка своего жилья                      620
строительство недвижимости                619
ремонт жилью                              607
покупка жилой недвижимости                606
на покупку своего автомобиля              505
заняться высшим образованием      

In [57]:
# Объявление функции purpose_category() с параметром purpose,
# которая возвращает следующие категории на основании полученого значения:
# 'операции с автомобилем',
# 'операции с недвижимостью',
# 'проведение свадьбы',
# 'получение образования'.

def purpose_category(purpose):
    purpose = str(purpose)
    if  'автомобил' in purpose:
        return 'операции с автомобилем'
    elif 'жиль' in purpose or 'недвижимост' in purpose:
        return 'операции с недвижимостью'
    elif 'свадьб' in purpose:
        return 'проведение свадьбы'
    elif 'образовани' in purpose:
        return 'получение образования'
    else:
        return 'без категории'    

In [58]:
df['purpose_category'] = df['purpose'].apply(purpose_category) # присвоение категорий по значением колонки total_income

In [59]:
df['purpose_category'].value_counts()

операции с недвижимостью    10811
операции с автомобилем       4306
получение образования        4013
проведение свадьбы           2324
Name: purpose_category, dtype: int64

## Ответы на вопросы.

##### Вопрос 1:

Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [60]:
round(df['debt'].mean(), 3) # доля заемщиков имеющих задолжности

0.081

In [61]:
# Объявление функции conclusion_pivot_table() с параметром data и column,
# которая возвращает датафрейм содержащий сводную таблицу по заданной колонке, со следующими колонками:
# 'процент задолженности по значениям заданной колонки',
# 'отклонение от среднего значения по всему датафрейму',
# 'доля записей по каждому значению во всем датафрейме'

def conclusion_pivot_table(data, column):
    debt_mean = round(data['debt'].mean(), 3) # доля заемщиков имеющих задолжности
    pivot = data.pivot_table( #сводная таблица по колонки, где в качестве значений рассчитываются доля заемщиков с задолженностями
        index=column,
        values='debt',
        aggfunc='mean')
    pivot['debt'] = round(pivot['debt'] * 100, 2) # перевод доли в проценты
    pivot['deviation'] = round((pivot['debt'] - 100 * debt_mean) / debt_mean, 2) # отклонений в процентах от среднего значения 
    pivot['records_share'] = round(data.groupby(column)[column].count() / data.shape[0], 3) #доля записей по колонке
    return(pivot)                      

In [62]:
conclusion_pivot_table(df, 'children') # вывод итоговой таблицы по количетву детей

debt  deviation  records_share
children                                
0         7.54      -6.91          0.657
1         9.17      13.21          0.226
2         9.49      17.16          0.099
3         8.18       0.99          0.015
4         9.76      20.49          0.002
5         0.00    -100.00          0.000

##### Вывод 1:

Зависимость между количеством детей и возвратом кредита в срок явна прослеживается, люди с большим количеством детей имеют больше задолжностей по возврату кредитов. При этом чем больше детей тем ситуация ухудшается, что логично т.к. дети это большие дополнительные расходы в том числе и непредвиденные. 
Можно заметить, что для трех и пяти детей вывод о растущей задолженности не подтверждается:
- для 3-х детей он находится на уровне среднего значения 8.2 % имеют задолженности в этой группе,
- для 5-и он вообще равен нулю,
Но также необходимо учесть, что процент заемщиков с 3, 4 и 5 детьми незначителен и составляет менее 2%, что скорее всего говорит о нерепрезентативности выборки в данных значениях.

##### Вопрос 2:

Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [63]:
df_family_status = df.merge(  # объединение основного датафрейма со словарем семейное положение 
    family_status_dict, 
    on='family_status_id', 
    how='left'
    )

In [64]:
conclusion_pivot_table(df_family_status, 'family_status') # вывод итоговой таблицы по семейному положению

debt  deviation  records_share
family_status                                        
в разводе              7.11     -12.22          0.056
вдовец / вдова         6.57     -18.89          0.045
гражданский брак       9.35      15.43          0.193
женат / замужем        7.55      -6.79          0.575
не женат / не замужем  9.75      20.37          0.131

##### Вывод 2:

Зависимость между семейным положением и возвратом кредита в срок явна прослеживается. Условно можно разделить на 2 категории: 
- заемщики, которые состоят или состояли в офицальном браке(в разводе и овдовевшие) луше справляются с возвращеним кредитов их показатели задолжностей ниже среднего,
- заемщики, которые не состоят в в офицальном браке (в том числе и гражданский брак), значительно хуже справляются с возвращеним кредитов

Разница между категорией `женат / замужем` и `не женат / не замужем` составляет более 27% 

##### Вопрос 3:

Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [65]:
conclusion_pivot_table(df, 'total_income_category') # вывод итоговой таблицы по уровню дохода

debt  deviation  records_share
total_income_category                                
B                      7.06     -12.84          0.235
C                      8.49       4.81          0.747
D                      6.00     -25.93          0.016
E                      9.09      12.22          0.001
А                      8.00      -1.23          0.001

##### Вывод 3:

К сожалению однозначный ответ на вопрос 3 по имеющимся данным сделать не получается по следующим причинам:
- на категории `B` и `C` приходятся более 98% всей выборки, а на оставшиеся категории приходятся менее 2% и оценить их репрезентативность очень сложно,
- 10% записей были заполнены медианными значениями, а это еще больше увеличило категории `C` (145000)
Если опереться на данные по категория `B` и `C` то чем выше доход тем ниже процент задолженности

##### Вопрос 4:

Как разные цели кредита влияют на его возврат в срок?

In [66]:
conclusion_pivot_table(df, 'purpose_category') # вывод итоговой таблицы по цели кредита

debt  deviation  records_share
purpose_category                                        
операции с автомобилем    9.36      15.56          0.201
операции с недвижимостью  7.23     -10.74          0.504
получение образования     9.22      13.83          0.187
проведение свадьбы        8.00      -1.23          0.108

##### Вывод 4:

Результат получился следующим:
- самый низкий показатель задолженности получился в категории `операции с недвижимостью`, что довольно предсказуемо, т.к это обычно очень взвешенное решение и банки зачастую серьезно подходят к одобрению таких кредитов,
- на уровне среднего значения находятся категория `проведение свадьбы`,
- две категории являются относительно рискованными `операции с автомобилем` и `получение образования`. Тут тоже все логично автомобиль помимо обслуживания кредита  - это дополнительные расходы на бензин, обслуживание, парковку, страховку и так далее. А получение образование связано с перераспределением времени и выделением его на образование следовательно уменьшение времени, которое человек может выделить на зарабатывание денег - это потеря дохода.

## Общий вывод:

In [68]:
df_family_status.pivot_table( # сводная таблица по зависимости задолженности от семейного положения и количества детей 
    index='family_status',
    columns='children',
    values= 'debt',
    aggfunc='mean')[[0, 1, 2]]

children                      0         1         2
family_status                                      
в разводе              0.070153  0.066456  0.096386
вдовец / вдова         0.062574  0.086420  0.125000
гражданский брак       0.083883  0.118000  0.092958
женат / замужем        0.069095  0.082224  0.093552
не женат / не замужем  0.092838  0.114537  0.119048

Были получены следующие результаты:
- Заемщики находящиеся/находившиеся в официальном браке и без детей являются самой привлекательной, с точки зрения кредитования, категорией.  Наличия одного (исключение для разведенных, скорее всего из-за малого числа данных)  и более ребенка сдвигает этот показатель на уровень выше среднего.
- Заемщики не состоящие в официальном браке даже без детей, имеют процент задолженности выше среднего, а появление детей еще сильнее усугубляет положение, и разница довольна значительна.